bounty_hunter/src/network_endpoints.rs

```diff
@@ -0,0 +1,92 @@
+use actix_web::{http::StatusCode, HttpRequest, HttpResponse, Json};
+use diesel::prelude::*;
+use failure::Error;
+use futures::{future, Future};
+use num256::Uint256;
+
+use std::collections::HashMap;
+
+use models::{ChannelState, ChannelStateRecord, NewChannelStateRecord};
+use schema::states::dsl::*;
+use DB_CONN;
+
+/// Pass channel state to the bounty hunter
+pub fn handle_upload_channel_state(
+    state_obj: Json<ChannelState>,
+) -> Box<Future<Item = HttpResponse, Error = Error>> {
+    trace!("Hit /upload_channel_state");
+
+    let mut ret = HashMap::<String, String>::new();
+
+    let state = state_obj.into_inner();
+    debug!("Got state {:?}", state);
+
+    match diesel::insert_into(states)
+        .values(NewChannelStateRecord::from(state))
+        .execute(&*DB_CONN.lock().unwrap())
+    {
+        Ok(_) => {}
+        Err(e) => {
+            ret.insert(
+                "error".to_owned(),
+                "Could not insert record into database".to_owned(),
+            );
+            ret.insert("rust_error".to_owned(), format!("{:?}", e));
+            return Box::new(future::ok(
+                HttpResponse::new(StatusCode::INTERNAL_SERVER_ERROR)
+                    .into_builder()
+                    .json(ret),
+            ));
+        }
+    }
+
+    return Box::new(future::ok(HttpResponse::Ok().json(ret)));
+}
+
+/// Query for the bounty hunter channel state from a requested time period
+pub fn handle_get_channel_state(
+    _req: HttpRequest,
+    ch_id: actix_web::Path<Uint256>,
+) -> Box<Future<Item = HttpResponse, Error = Error>> {
+    let ch_id = ch_id.into_inner();
+    trace!("Hit /get_channel_state/{}", ch_id);
+
+    let matching_records = match states
+        .filter(channel_id.eq(format!("0x{}", ch_id)))
+        .load::<ChannelStateRecord>(&*DB_CONN.lock().unwrap())
+    {
+        Ok(values) => values,
+        Err(e) => {
+            let mut err_ret = HashMap::<String, String>::new();
+            err_ret.insert(
+                "error".to_owned(),
+                "Could not retrieve records from database".to_owned(),
+            );
+            err_ret.insert("rust_error".to_owned(), format!("{:?}", e));
+            return Box::new(future::ok(
+                HttpResponse::new(StatusCode::INTERNAL_SERVER_ERROR)
+                    .into_builder()
+                    .json(err_ret),
+            ));
+        }
+    };
+
+    info!(
+        "Found {} records for channel id {}",
+        matching_records.len(),
+        ch_id
+    );
+
+    debug!("Records: {:?}", matching_records);
+
+    let mut ok_ret = HashMap::<String, Vec<ChannelState>>::new();
+    ok_ret.insert(
+        "records".to_owned(),
+        matching_records
+            .iter()
+            .map(|record| ChannelState::from_record(record.clone()).unwrap())
```

Comment:
Just curious - wouldn't `move |record| ChannelState::from_record(record)` work without cloning?

Reply:
Good idea, will check

Reply:
`into_iter` did it and consumed the `matched_records` vec, good point though. BH is supposed to process a lot of data at once, a stream-like approach like this is always welcome.